In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as mlt
import seaborn as sns

In [ ]:
df=pd.read_excel('data/worldwide_governance.xlsx')
df.head()

,Country Name,Country Code,Series Name,Series Code,2015 [YR2015],2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022],2023 [YR2023],2024 [YR2024]
0,Australia,AUS,Government Effectiveness - Governance estimate...,GOV_WGI_GE.EST,1.676215,1.624845,1.644298,1.724781,1.666649,1.605457,1.702487,1.767515,1.781813,1.809045
1,Australia,AUS,Government Effectiveness - Governance score (0...,GOV_WGI_GE.SC,83.932833,82.934005,83.312249,84.877164,83.746839,82.557009,84.443678,85.708068,85.986094,86.515590
2,Australia,AUS,Political Stability - Governance estimate (app...,GOV_WGI_PV.EST,0.803247,0.954078,0.783148,0.884094,0.845414,0.885940,0.853651,0.901535,0.925436,0.793269
3,Australia,AUS,Political Stability - Governance score (0-100),GOV_WGI_PV.SC,79.615144,82.191581,79.271813,80.996137,80.335419,81.027661,80.476113,81.294059,81.702330,79.444702
4,Australia,AUS,Rule of Law - Governance estimate (approx. -2....,GOV_WGI_RL.EST,1.717785,1.728829,1.623700,1.635892,1.634261,1.552139,1.570184,1.471946,1.503184,1.470401


In [ ]:
df.shape

(173, 14)

In [ ]:
df.columns

Index(['Country Name', 'Country Code', 'Series Name', 'Series Code',
       '2015 [YR2015]', '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]',
       '2019 [YR2019]', '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]',
       '2023 [YR2023]', '2024 [YR2024]'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173 entries, 0 to 172
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Country Name   170 non-null    object 
 1   Country Code   168 non-null    object 
 2   Series Name    168 non-null    object 
 3   Series Code    168 non-null    object 
 4   2015 [YR2015]  168 non-null    float64
 5   2016 [YR2016]  168 non-null    float64
 6   2017 [YR2017]  168 non-null    float64
 7   2018 [YR2018]  168 non-null    float64
 8   2019 [YR2019]  168 non-null    float64
 9   2020 [YR2020]  168 non-null    float64
 10  2021 [YR2021]  168 non-null    float64
 11  2022 [YR2022]  168 non-null    float64
 12  2023 [YR2023]  168 non-null    float64
 13  2024 [YR2024]  168 non-null    float64
dtypes: float64(10), object(4)
memory usage: 19.1+ KB


clean the data


1.modify column name

In [ ]:
df.columns

Index(['Country Name', 'Country Code', 'Series Name', 'Series Code',
       '2015 [YR2015]', '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]',
       '2019 [YR2019]', '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]',
       '2023 [YR2023]', '2024 [YR2024]'],
      dtype='object')

In [ ]:
df.columns = df.columns.str.replace(r"\[YR\d+\]","",regex=True).str.strip()
df.columns

Index(['Country Name', 'Country Code', 'Series Name', 'Series Code', '2015',
       '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024'],
      dtype='object')

Transform data

Melt df 

In [ ]:
year_col=[col for col in df.columns if col.isdigit()]
year_col

['2015',
 '2016',
 '2017',
 '2018',
 '2019',
 '2020',
 '2021',
 '2022',
 '2023',
 '2024']

In [ ]:
df_melted=df.melt(id_vars=['Country Name','Country Code','Series Name'],
                  value_vars=year_col,
                  var_name="year",
                  value_name="Value")

In [ ]:
df_melted.head()

,Country Name,Country Code,Series Name,year,Value
0,Australia,AUS,Government Effectiveness - Governance estimate...,2015,1.676215
1,Australia,AUS,Government Effectiveness - Governance score (0...,2015,83.932833
2,Australia,AUS,Political Stability - Governance estimate (app...,2015,0.803247
3,Australia,AUS,Political Stability - Governance score (0-100),2015,79.615144
4,Australia,AUS,Rule of Law - Governance estimate (approx. -2....,2015,1.717785


In [ ]:
df_melted.shape

(1730, 5)

clean missing data

In [ ]:
df_melted.isnull().sum()

Country Name    30
Country Code    50
Series Name     50
year             0
Value           50
dtype: int64

50/1730=2.8%
less than 5% 
so we'll drop the missing rows in the data

In [ ]:
df_melted[df_melted.isnull().any(axis=1)].head()

,Country Name,Country Code,Series Name,year,Value
168,NaN,NaN,NaN,2015,NaN
169,NaN,NaN,NaN,2015,NaN
170,NaN,NaN,NaN,2015,NaN
171,Data from database: Worldwide Governance Indic...,NaN,NaN,2015,NaN
172,Last Updated: 03/18/2026,NaN,NaN,2015,NaN


it is required these rows be removed since not significant for any analysis

In [ ]:
df_melted = df_melted[
    df_melted["Country Name"].notna()
]

In [ ]:
df_melted.isnull().sum()

Country Name     0
Country Code    20
Series Name     20
year             0
Value           20
dtype: int64

In [ ]:
df_melted = df_melted[
    df_melted["Series Name"].notna()
]

In [ ]:
df_melted.isnull().sum()

Country Name    0
Country Code    0
Series Name     0
year            0
Value           0
dtype: int64

In [ ]:
df_melted[df_melted.isnull().any(axis=1)].head()

,Country Name,Country Code,Series Name,year,Value


In [ ]:
df_melted.shape

(1680, 5)

Transform Series Name column

In [ ]:
df_melted['Series Name']

0       Government Effectiveness - Governance estimate...
1       Government Effectiveness - Governance score (0...
2       Political Stability - Governance estimate (app...
3          Political Stability - Governance score (0-100)
4       Rule of Law - Governance estimate (approx. -2....
                              ...                        
1720    Voice and Accountability - Governance score (0...
1721         Voice and Accountability - Number of sources
1722                      Rule of Law - Number of sources
1723              Political Stability - Number of sources
1724         Government Effectiveness - Number of sources
Name: Series Name, Length: 1680, dtype: object

In [ ]:
pd.set_option('display.max_colwidth',None)

In [ ]:
df_melted['Series Name'].head(10)

0    Government Effectiveness - Governance estimate (approx. -2.5 to +2.5)
1                      Government Effectiveness - Governance score (0-100)
2         Political Stability - Governance estimate (approx. -2.5 to +2.5)
3                           Political Stability - Governance score (0-100)
4                 Rule of Law - Governance estimate (approx. -2.5 to +2.5)
5                                   Rule of Law - Governance score (0-100)
6    Voice and Accountability - Governance estimate (approx. -2.5 to +2.5)
7                      Voice and Accountability - Governance score (0-100)
8                             Voice and Accountability - Number of sources
9                                          Rule of Law - Number of sources
Name: Series Name, dtype: object

In [ ]:
df_melted[['Metric','Type']] = df_melted['Series Name'].str.split(" - ",expand=True)

In [ ]:
df_melted[['Metric','Type']] 

,Metric,Type
0,Government Effectiveness,Governance estimate (approx. -2.5 to +2.5)
1,Government Effectiveness,Governance score (0-100)
2,Political Stability,Governance estimate (approx. -2.5 to +2.5)
3,Political Stability,Governance score (0-100)
4,Rule of Law,Governance estimate (approx. -2.5 to +2.5)
...,...,...
1720,Voice and Accountability,Governance score (0-100)
1721,Voice and Accountability,Number of sources
1722,Rule of Law,Number of sources
1723,Political Stability,Number of sources


In [ ]:
df_melted['Metric']=df_melted['Metric'].str.strip()
df_melted['Type']=df_melted['Type'].str.strip()

In [ ]:
df_melted[['Metric','Type']] 

,Metric,Type
0,Government Effectiveness,Governance estimate (approx. -2.5 to +2.5)
1,Government Effectiveness,Governance score (0-100)
2,Political Stability,Governance estimate (approx. -2.5 to +2.5)
3,Political Stability,Governance score (0-100)
4,Rule of Law,Governance estimate (approx. -2.5 to +2.5)
...,...,...
1720,Voice and Accountability,Governance score (0-100)
1721,Voice and Accountability,Number of sources
1722,Rule of Law,Number of sources
1723,Political Stability,Number of sources


In [ ]:
df_melted['Metric'].unique()

array(['Government Effectiveness', 'Political Stability', 'Rule of Law',
       'Voice and Accountability'], dtype=object)

In [ ]:
df_melted['Type'].unique()

array(['Governance estimate (approx. -2.5 to +2.5)',
       'Governance score (0-100)', 'Number of sources'], dtype=object)

In [ ]:
df_melted['Type'].value_counts()

Type
Governance estimate (approx. -2.5 to +2.5)    560
Governance score (0-100)                      560
Number of sources                             560
Name: count, dtype: int64

In [ ]:
df_melted[df_melted["Type"].str.contains("score", case=False)]["Metric"].value_counts()

Metric
Government Effectiveness    140
Political Stability         140
Rule of Law                 140
Voice and Accountability    140
Name: count, dtype: int64

In [ ]:
140*4

560

Removing “Governance Estimate (-2.5 to +2.5)”
->Governance estimates are measured on a different scale (-2.5 to +2.5), making them less interpretable and not directly comparable with normalized scores

Removing “Number of Sources”
->The number of sources reflects data reliability rather than governance performance, so it was excluded from analytical comparisons.

To ensure consistency and meaningful comparisons,filtered the dataset to include only governance scores (0–100), as other types either used different scales or represented metadata rather than actual performance indicators.

Changes in df_final will NOT affect df_melted

To ensure that the filtered dataset is independent and to avoid unintended side effects or chained assignment issues during further transformations.

In [ ]:
df_final=df_melted[df_melted['Type'].str.contains("score", case=False)].copy()

In [ ]:
df_final['Type'].value_counts()

Type
Governance score (0-100)    560
Name: count, dtype: int64

In [ ]:
df_final.head()

,Country Name,Country Code,Series Name,year,Value,Metric,Type
1,Australia,AUS,Government Effectiveness - Governance score (0-100),2015,83.932833,Government Effectiveness,Governance score (0-100)
3,Australia,AUS,Political Stability - Governance score (0-100),2015,79.615144,Political Stability,Governance score (0-100)
5,Australia,AUS,Rule of Law - Governance score (0-100),2015,86.529000,Rule of Law,Governance score (0-100)
7,Australia,AUS,Voice and Accountability - Governance score (0-100),2015,84.057637,Voice and Accountability,Governance score (0-100)
13,China,CHN,Government Effectiveness - Governance score (0-100),2015,61.716004,Government Effectiveness,Governance score (0-100)


KPI ANALYSIS

1. average governance score (global)

    Overall governance level worldwide

In [ ]:
global_avg=df_final['Value'].mean()
global_avg

np.float64(63.65202313071428)

2. india governance score

In [ ]:
df_india=df_final[df_final['Country Code']=='IND']
india_avg=df_india['Value'].mean()
india_avg

np.float64(56.664567415)

In [ ]:
india_trend=df_india.groupby("year")["Value"].mean()
india_trend

year
2015    57.147546
2016    57.425220
2017    57.979614
2018    58.279040
2019    58.095412
2020    56.188031
2021    55.149033
2022    55.288183
2023    55.234275
2024    55.859319
Name: Value, dtype: float64

3. top and bottom countried

In [ ]:
latest_year=df_final['year'].max()
df_latest=df_final[df_final['year']==latest_year]
df_latest.head()

,Country Name,Country Code,Series Name,year,Value,Metric,Type
1558,Australia,AUS,Government Effectiveness - Governance score (0-100),2024,86.515590,Government Effectiveness,Governance score (0-100)
1560,Australia,AUS,Political Stability - Governance score (0-100),2024,79.444702,Political Stability,Governance score (0-100)
1562,Australia,AUS,Rule of Law - Governance score (0-100),2024,82.271353,Rule of Law,Governance score (0-100)
1564,Australia,AUS,Voice and Accountability - Governance score (0-100),2024,83.318782,Voice and Accountability,Governance score (0-100)
1570,China,CHN,Government Effectiveness - Governance score (0-100),2024,68.758124,Government Effectiveness,Governance score (0-100)


In [ ]:
df_final['Country Name'].value_counts()

Country Name
Australia               40
China                   40
Japan                   40
India                   40
Russian Federation      40
United Kingdom          40
United States           40
United Arab Emirates    40
Afghanistan             40
Germany                 40
New Zealand             40
Philippines             40
South Africa            40
Ukraine                 40
Name: count, dtype: int64

out of 14 countries, let's find top 5 countries based on governance

In [ ]:
top_countries=df_latest.groupby("Country Name")['Value'].mean().sort_values(ascending=False).head(5)
top_countries

Country Name
New Zealand       86.230828
Japan             84.143125
Australia         82.887607
Germany           79.797575
United Kingdom    75.777590
Name: Value, dtype: float64

lets see the order of them

In [ ]:
top_countries_order=df_latest.groupby("Country Name")['Value'].mean().sort_values(ascending=False).head(14)
top_countries_order

Country Name
New Zealand             86.230828
Japan                   84.143125
Australia               82.887607
Germany                 79.797575
United Kingdom          75.777590
United States           71.986152
United Arab Emirates    65.656924
South Africa            56.762023
India                   55.859319
China                   51.735435
Philippines             51.133951
Ukraine                 49.420554
Russian Federation      41.142716
Afghanistan             21.215647
Name: Value, dtype: float64

4. metric-wise analysis
which governance pillar is strongest

In [ ]:
metric_avg=df_final.groupby('Metric')['Value'].mean().sort_values(ascending=False)
metric_avg

Metric
Government Effectiveness    65.466448
Political Stability         64.214486
Rule of Law                 63.093552
Voice and Accountability    61.833607
Name: Value, dtype: float64

5. trend over time (global)

In [ ]:
global_trend=df_final.groupby('year')['Value'].mean()
global_trend

year
2015    64.730846
2016    64.463637
2017    64.354438
2018    64.382078
2019    64.354102
2020    63.682069
2021    63.115274
2022    62.547774
2023    62.479337
2024    62.410675
Name: Value, dtype: float64

6. stability analysis
which countries are stable vs volatile

In [ ]:
stability=df_final.groupby('Country Name')['Value'].std().sort_values()
stable_countries=stability.head(7)
unstable_countries=stability.tail(7)

low std -> value don't change much so stable

high std -> value fluctuate a lot so unstable

In [ ]:
stable_countries

Country Name
Australia         1.973143
New Zealand       2.077045
India             2.789356
United Kingdom    4.068715
Philippines       4.176987
Germany           4.422994
Japan             4.446303
Name: Value, dtype: float64

In [ ]:
unstable_countries

Country Name
South Africa             5.598676
United States            5.833303
Afghanistan              6.521273
Ukraine                  6.719654
Russian Federation       6.896638
China                   13.580167
United Arab Emirates    14.605453
Name: Value, dtype: float64

In [ ]:
country_stats = df_final.groupby("Country Name")["Value"].agg(["mean", "std"]).sort_values("mean", ascending=False)
country_stats.head(14)

,mean,std
Country Name,,
New Zealand,87.396284,2.077045
Australia,83.226691,1.973143
Germany,83.194252,4.422994
Japan,83.098324,4.446303
United Kingdom,78.486773,4.068715
United States,74.747208,5.833303
United Arab Emirates,65.704884,14.605453
South Africa,58.650662,5.598676
India,56.664567,2.789356


Top Performers (High Mean + Low Std)

1.strong & Stable Countries

New Zealand → (87.3, 2.07)

Australia → (83.2, 1.97)


These countries show consistently high governance performance with minimal fluctuations, indicating strong and stable institutional systems.

2.Strong but Slightly Variable
Germany → high mean, moderate std

Japan → high mean, moderate std

United Kingdom → good mean, moderate std


These countries maintain strong governance performance but exhibit moderate variability, possibly due to policy changes or external factors.

3.Moderate but Unstable

United States → (74.7, 5.8)

United Arab Emirates → (65.7, 14.6)  --> high std value


While maintaining relatively strong governance scores, these countries show higher fluctuations, indicating inconsistent performance across years.

4.Mid-Level Performers

India → (56.6, 2.78)  stable

South Africa → (58.6, 5.59)  unstable


India demonstrates stable but moderate governance performance, whereas South Africa shows similar performance levels with higher variability.

5.Weak & Unstable Countries

China → (50.5, 13.5) 

Ukraine → (47.5, 6.7)

Russia → (45.8, 6.8)

Afghanistan → (25.1, 6.5)


These countries exhibit lower governance scores along with high variability, indicating structural challenges and instability.

----------------------------------------------
The analysis shows that high-performing countries tend to exhibit greater stability, while lower-performing countries often display higher variability, highlighting a correlation between governance strength and consistency.

----------------------------------------------------------------------
Category	               ******             Condition

Strong & Stable	           ******             High mean, Low std

Strong but Variable	           ******         High mean, High std

Moderate Stable	             ******            Medium mean, Low std

Weak & Unstable	             ******           Low mean, High std

I evaluated governance performance using both average scores and variability. This allowed me to distinguish between consistently strong countries and those with fluctuating governance trends, providing deeper insights beyond simple rankings.

forgot to drop the series name column from df_final 

since it already got metric and type

also drop type since we only focused on "score(0-100)"

In [ ]:
df_final=df_final.drop(columns=['Series Name','Type'])

In [ ]:
df_final.to_csv("data/governance_cleaned.csv", index=False)